pandoc.bzl

```diff
@@ -1,90 +1,104 @@
+FORMAT_EXTENSIONS = {
+    "asciidoc": "adoc",
+    "beamer": "tex",
+    "commonmark": "md",
+    "context": "tex",
+    "docbook": "xml",
+    "docbook4": "xml",
+    "docbook5": "xml",
+    "docx": "docx",
+    "dokuwiki": "txt",
+    "dzslides": "html",
+    "epub": "epub",
+    "epub2": "epub",
+    "epub3": "epub",
+    "fb2": "fb",
+    "haddock": "txt",
+    "html": "html",
+    "html4": "html",
+    "html5": "html",
+    "icml": "icml",
+    "jats": "xml",
+    "json": "json",
+    "latex": "tex",
+    "man": "1",
+    "markdown": "md",
+    "markdown_github": "md",
+    "markdown_mmd": "md",
+    "markdown_phpextra": "md",
+    "markdown_strict": "md",
+    "mediawiki": "txt",
+    "ms": "1",
+    "muse": "txt",
+    "native": "txt",
+    "odt": "odt",
+    "opendocument": "odt",
+    "opml": "openml",
+    "org": "txt",
+    "plain": "txt",
+    "pptx": "pptx",
+    "revealjs": "html",
+    "rst": "rst",
+    "rtf": "rtf",
+    "s5": "html",
+    "slideous": "html",
+    "slidy": "html",
+    "tei": "html",
+    "texinfo": "texi",
+    "textile": "textile",
+    "zimwiki": "txt",
+}
+PANDOC_FORMATS = FORMAT_EXTENSIONS.keys()
+
 def _pandoc_impl(ctx):
     toolchain = ctx.toolchains["@bazel_pandoc//:pandoc_toolchain_type"]
+    cli_args = ctx.attr.options + [
+        "--from",
+        ctx.attr.from_format,
+        "--to",
+        ctx.attr.to_format,
+        "-o",
+        ctx.outputs.output.path,
+        ctx.file.src.path,
+    ]
+    # print("args=" + str(cli_args))
     ctx.actions.run(
         mnemonic = "Pandoc",
         executable = toolchain.pandoc.files.to_list()[0].path,
-        arguments = ctx.attr.options + [
-            "--from",
-            ctx.attr.from_format,
-            "--to",
-            ctx.attr.to_format,
-            "-o",
-            ctx.outputs.out.path,
-            ctx.files.src[0].path,
-        ],
+        arguments = cli_args,
         inputs = toolchain.pandoc.files + ctx.files.src,
-        outputs = [ctx.outputs.out],
+        outputs = [ctx.outputs.output],
     )
 
-_pandoc = rule(
+pandoc_rule = rule(
     attrs = {
         "extension": attr.string(),
-        "from_format": attr.string(),
+        "from_format": attr.string(mandatory = True),
         "options": attr.string_list(),
-        "src": attr.label(allow_files = True),
-        "to_format": attr.string(),
+        "src": attr.label(allow_single_file = True, mandatory = True),
+        "to_format": attr.string(mandatory = True),
+        "output": attr.output(),
```

Comment:
I was thinking. If we're making the output file configurable, maybe we could simply remove `from_format` and `to_format` entirely. Pandoc seems to automatically infer the intended source/destination format based on the file extension. If people need something custom, they could pass that in through `options`. What are your thoughts on this?

This would allow us to get rid of `FORMAT_EXTENSIONS` and `PANDOC_FORMATS`.

Reply:
I didn't know pandoc can infer the input & output formats.
I'm happy to remove the list of extensions, yes!
There is a little benefit in having the extensions in the bazel rule: we can fail early if a format is not specified.

The thing is: output is still not mandatory (my change is backward compatible).
Would you like to make output mandatory, then?

Reply:
Exactly! Making the output mandatory sounds good in that case.

Reply:
Sure, will do. But we still need the formats for the test suite. So let's start by moving the formats in a common place (#3).
We can also move the samples in their own package (#2)

Reply:
On second thought, removing `--from_format` would be a breaking change for the users of the rule. I'll make it optional and keep the existing behaviour. Feel free to remove really it in a subsequent change.
Also, if the rule knows the format, bazel can do an early check at analysis phase, rather than invoking pandoc which will then fail.